pallene/coder.lua

```diff
@@ -462,6 +462,15 @@ function Coder:pallene_entry_point_definition(f_id)
         table.insert(prologue, decl..initializer..";"..comment)
     end
 
+    for u_id, upval in ipairs(func.captured_vars) do
+        local typ, c_name = self:prepare_captured_var(u_id, upval.decl)
+        local decl = C.declaration(ctype(typ), c_name)..";";
+        local src  = string.format("&U[%d]", u_id)
+        local init = unchecked_get_slot(typ, c_name, src)..C.comment(upval.decl.name)
+        table.insert(prologue, decl);
+        table.insert(prologue, init);
+    end
```

Comment:
Instead of eagerly fetching all the upvalues at the start of the function, could we instead fetch them lazily, only when we use them? This is what we currently do with the stuff in the `G`. See the `ir.Value.Function` and the `GetGlobal` for example.

If we say that the `U` is a `restrict` pointer, I think the C compiler should be able to optimize it to only fetch the upvalue once, even if we are fetching it "lazily".

before:

    Udata* uv1 = uvalue(&U[1]);
    int uv2 = ivalue(&U[2]);
    stuff(uv1, uv2);

after:

    stuff(uvalue(&U[1]), ivalue(&U[2]));

Reply:
Is the C compiler guaranteed to optimize this? (I'm not well aware :p)
If not, then maybe we could use a benchmark.

Reply:
I agree that some experiments wouldn't hurt. It might be more direct to look at the generated assembly instead of a performance benchmark, though. In #413 I link to some [tests](https://godbolt.org/z/zPhTja9sG) I ran on a simplified non-Pallene program. It suggested that, with restrict, Clang could optimize something similar to this but gcc did not.

That said, my initial instinct is to be optimistic and try to leave the optimization to C compiler. I also get a bit nervous creating all those extra variables; they add pressure to the register allocator. One thing that I like about the new calling convention is that it doesn't have as many variables (although admitedly I don't have hard performance data about that).

Reply:
Sounds good enough to me!
I don't know the internals of C compilers to make a more educated guess. I couldn't have figured out the problem with the register allocator you mentioned, so I'll believe this works the way we think it does.

Reply:
This might require some changes, since we'll have to know the type of an upvalue id. Would it be ok to maintain a map of `upvalue_id -> type` for each `ir.Function`?

Reply:
The only thing you need to know about the register allocation, is that in assembly language there are a finite number of CPU registers. If two C variables are "alive" at the same time, then they can't be stored in the same CPU register. If at some point in the program there are too many variables "alive", some have to be stored in the stack instead of in registers (spilling).

From this point of view, it's not ideal to initialize a lots of variables at the start of the function, only to use their value much later. That results in many variables with overlapping lifetimes. If we decide that we can't trust the C compiler to optimize the upvalues, I expect that it would be better to create the local variables closer to the point where they will be used, instead of at the start of the function. But that's more complicated than doing it the lazy way ;)

Reply:
> This might require some changes, since we'll have to know the type of an upvalue id

I expect that it would work similar to what we're already doing. I think we're getting the type from the `UpvalInfo` (captured_vars)?

Reply:
The reason I think we may need a map is because upvalues can be referenced in all kinds of places with `ir.Value.Upvalue(id)`, so for a given upvalue ID, we'd want to know what type it is and do something similar to `c_value` by casting it to the correct type (from U[id]).

Reply:
Exactly. But I think we already have that map? As I said before, I think it is the `captured_vars`.

Reply:
Yeah, that's what I meant xD
Now I see what caused the confusion. I should have phrased it better.

Reply:
Would it be okay to introduce a second parameter to `c_value` for this purpose (`func`)? That way c_value can pass that to c_upval and c_upval and return the appropriately casted string.

Reply:
I'd hate to need to pass an extra func parameter to the hundreds of places where we call c_value. Maybe we should store that in the `self`, as an instance variable of the Coder class.

Currently, we're operating under the assumption that all the upvalues are living in the `G`, and that all functions in the module have the same upvalues. That's why the `upvalue_of_foo` instance variables are defined at the module level... But now that each function has a different set of upvalues, I think it makes sense to keep a reference to the current function as an instance variable...

In a later PR we can refactor the places that currently receive the `func` as a parameter, and also switch them to using an instance variable.